src/parser.cpp

```diff
@@ -1332,14 +1333,21 @@ const Statement *Parser::parseVarStatement(Scope *scope, int line)
         } else {
             v = new LocalVariable(name, vars.second, scope, false);
         }
-        scope->addName(name, v);
+        variables.insert(std::map<std::string, Variable *>::value_type(name, v));
+        
         varrefs.push_back(new VariableExpression(v));
     }
     if (tokens[i].type == ASSIGN) {
         ++i;
         const Expression *expr = parseExpression(scope);
+        for (auto v: variables) {
+            scope->addName(v.first, v.second, true);
+        }
         return new AssignmentStatement(line, varrefs, expr);
     }
+    for (auto v: variables) {
+        scope->addName(v.first, v.second);
+    }
     return nullptr;
```

Comment:
I'm a bit uncomfortable with this logic for managing the return value. I might suggest something like:

    const Statement *r = nullptr;
    if (tokens[i].type == ASSIGN) {
        // assign stuff, set r to an AssignmentStatement
    } else {
        // normal addName without setting the referenced bit
    }
    return r;